userspace/libsinsp/event.cpp

```diff
@@ -1197,33 +1197,37 @@ Json::Value sinsp_evt::get_param_as_json(uint32_t id, OUT const char** resolved_
 			const struct ppm_name_value *flags = (const struct ppm_name_value *)m_info->params[id].info;
 			uint32_t initial_val = val;
 
-			while(flags != NULL && flags->name != NULL && flags->value != initial_val)
+			while(flags != NULL && flags->name != NULL)
 			{
-				if ((exact_match && flags->value == initial_val) ||
+				bool match = false;
+				if (exact_match)
+				{
+					match = flags->value == initial_val;
+				}
+				else
+				{
 					// If flag is 0, then initial_val needs to be 0 for the flag to be resolved
-					((flags->value == 0 && initial_val == 0) ||
-				     (flags->value != 0 && (val & flags->value) == flags->value && val != 0)))
+					if ((flags->value == 0 && initial_val == 0) ||
+					   (flags->value != 0 && (val & flags->value) == flags->value && val != 0))
+					{
+						match = true;
+						// We remove current flags value to avoid duplicate flags e.g. PPM_O_RDWR, PPM_O_RDONLY, PPM_O_WRONLY
+						val &= ~flags->value;
+					}
+				}
+				if (match)
 				{
 					ret["flags"].append(flags->name);
-
-					if (exact_match)
+					if (!exact_match && flags->value == initial_val)
 					{
-						// We found the match!
+						// if we reached initial val, we have finished.
+						// NOTE: for enum flags, we might have multiple flags matching same enum value
+						// see socket_families (eg: AF_LOCAL, AF_UNIX). Don't break.
```

Comment:
I don't understand. Do we want to return "AF_LOCAL AF_UNIX" in that case? Why?

Reply:
Because when a single enum value is matched by multiple flags, we want the user to be able to filter by:
```
evt.type=socket and evt.arg.domain contains AF_UNIX"
```
Otherwise, since `AF_LOCAL` is the first to appear, you would not be able to trigger the filter by matching `AF_UNIX`, even if they share the same value.

This is the same behavior we enforce with bitmask flags, ie:
```
evt.type=openat and evt.arg.flags contains O_CLOEXEC
```

But for bitmask flags we encode the string like `O_CLOEXEC|O_RDONLY` instead (notice `|` in place of the ` `)

Reply:
Is this ok now? :)